Cubical/Algebra/Group/Abelianization/AbelianizationAsCoeq.agda

```diff
@@ -0,0 +1,398 @@
+      (λ x → cong incAb (rid x))
+
+  rinvAb : (x : Abelianization) → x ·Ab (invAb x) ≡ 1Ab
+  rinvAb =
+    elimProp
+      (λ x → squash (x ·Ab (invAb x)) 1Ab)
+      (λ x → (incAb x) ·Ab (invAb (incAb x)) ≡⟨ refl ⟩
+             (incAb x) ·Ab (incAb (inv x))   ≡⟨ refl ⟩
+             incAb (x · (inv x))             ≡⟨ cong incAb (fst (inverse x)) ⟩
+             incAb 1g                        ≡⟨ refl ⟩
+             1Ab ∎)
+
+  commAb : (x y : Abelianization) → x ·Ab y ≡ y ·Ab x
+  commAb =
+    elimProp2
+      (λ x y → squash (x ·Ab y) (y ·Ab x))
+      (λ x y → (incAb x) ·Ab (incAb y) ≡⟨ refl ⟩
+               incAb (x · y)           ≡⟨ cong incAb (sym (lid (x · y))) ⟩
+               incAb (1g · (x · y))    ≡⟨ comm 1g x y ⟩
+               incAb (1g · (y · x))    ≡⟨ cong incAb (lid (y · x)) ⟩
+               incAb (y · x)           ≡⟨ refl ⟩
+               (incAb y) ·Ab (incAb x) ∎)
+
+  -- The proof that the abelianization is in fact an abelian group.
+  asAbelianGroup : AbGroup ℓ
+  asAbelianGroup = makeAbGroup 1Ab _·Ab_ invAb squash assocAb ridAb rinvAb commAb
+
+  -- The proof that incAb can be seen as a group homomorphism
+  incAbAsGroupHom : GroupHom G (AbGroup→Group asAbelianGroup)
+  incAbAsGroupHom = f , fIsHom
+    where
+    f = λ x → incAb x
+    fIsHom : IsGroupHom (snd G) f (snd (AbGroup→Group asAbelianGroup))
+    IsGroupHom.pres· fIsHom = λ x y → refl
+    IsGroupHom.pres1 fIsHom = refl
+    IsGroupHom.presinv fIsHom = λ x → refl
+
+AbelianizationAbGroup : (G : Group ℓ) → AbGroup ℓ
+AbelianizationAbGroup G = AbelianizationAsCoeq.asAbelianGroup G
+
+AbelianizationHom : (G : Group ℓ) → GroupHom G (AbGroup→Group (AbelianizationAbGroup G))
+AbelianizationHom G = AbelianizationAsCoeq.incAbAsGroupHom G
+
+module UniversalPropertyCoeq (G : Group ℓ) where
+  open GroupStr {{...}}
+  open GroupTheory G
+  open AbelianizationAsCoeq G
+  private
+    instance
+      _ = snd G
+    G' = fst G
+  abstract
+    {- The proof of the universal property of the abelianization.
+
+    G -incAb-> Abelianization
+    \         .
+      \       .
+        f   ∃! inducedHom
+          \   .
+            \ .
+              H
+    commuting diagram
+    -}
+    inducedHom : (H : AbGroup ℓ)
+              → (f : GroupHom G (AbGroup→Group H))
+              → AbGroupHom asAbelianGroup H
+    inducedHom H f = g , gIsHom
+      where open IsGroupHom
+            instance
+              _ : GroupStr (fst H)
+              _ = snd (AbGroup→Group H)
+            f' : fst G → fst H
+            f' = fst f
+            g : Abelianization → fst H
+            g = rec
+                  (isSetAbGroup H)
+                  (λ x → (f') x)
+                  (λ (a , b , c) → f' (a · b · c)     ≡⟨ (snd f).pres· a (b · c) ⟩
+                            (f' a) · (f' (b · c))    ≡⟨ cong (λ x → (f' a) · x) ((snd f).pres· b c) ⟩
+                            (f' a) · (f' b) · (f' c) ≡⟨ cong (λ x → (f' a) · x) ((snd H).AbGroupStr.comm (f' b) (f' c)) ⟩
```

Comment:
Please line break here... (you can just put "((snd ..." under "(lambda x -> ..." )

Reply:
done